_posts/2021/2021-09-28-cpp-lessons-learned.md

```diff
@@ -0,0 +1,143 @@
+
+Since we keep all data for our application on persistent memory, one might think we should
+do that as well with locks. That's the wrong way. Each lock and unlock operation might end up
+with write to pmem. We measured the performance, and it turned out, this approach is slow
+and inefficient. There are two solutions we propose to fix this problem:
+
+The first one is to keep per-element locks in DRAM instead (pointers to these locks are held
+on pmem). Currently, within our library, it is only possible with `concurrent_hash_map`.
+It's a good approach if you store large elements.
+
+The second solution is to use sharding instead of per-element locks. Data can be divided
+into multiple single-threaded buckets, each protected by a lock. With a uniform distribution
+of elements and a big enough number of buckets, this solution will provide a short waiting
+time on locks and good performance. In practice, this can be useful, most likely only in
+unsorted data structures.
+
+
+* **Lazy rehashing**
+
+Usually, in a structure with hashing table, it is required to perform, from time to time,
+a full rehashing of existing entries. In a concurrent environment, this task is very
+challenging and may require locking the entire structure. In our `concurrent_hash_map`
+we introduced a "lazy rehashing" solution, based on the paper
+["Per-bucket concurrent rehashing algorithms"][rehashing_pdf]. When rehashing is required,
+the number of buckets is doubled - some elements may be moved into new buckets, and some
+may stay in the old ones. No locks on buckets are taken, so the whole process happens "lazily"
+in the background. All the algorithm's details can be found in the linked article.
+In short - we use more memory but gain the performance. **Without this approach**,
+the structure (or rather its performance) is vulnerable in moments of mentioned
+full rehashing. These time slots cause latency spikes, slowing down each new write to
+the structure (because of a lock on the whole data structure).
+
+
+* **Layout compatibility**
+
+By publishing a stable release of libpmemobj-cpp, we committed to a stable on-media layout.
+From that point, we try and still promise that. We had a small hiccup on the way (in PMDK)
+with pools compatibility, but we've managed to fix that with the help of a tool now called
+[pmdk-convert][pmdk_convert_blog]. Right now, we've learned our lesson, and we try to predict
+as much as possible in the layout, but we know that *we don't know everything*. That's why
+keeping in mind future improvements we save up some space for extensions in each
+structure. We've also introduced a mechanism for checking the compatibility - a `feature_flags`.
+Each time a data structure, which implements feature flags, is used we check
+for compatibility. For example, in `concurrent_hash_map` on each initialization, we call
+`check_incompat_features()` method, and it throws an exception in case of flags incompatibility.
```

Comment:
I think you might extend this description to mention what kind of information we keep in those flags. For example, size of key/value pair, support for certain features, etc.